flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+		int deleteCount = 0;
+		long prevNode = findPredecessor(keyByteBuffer, keyOffset, 1);
+		long currentNode  = helpGetNextNode(prevNode, 0);
+		long nextNode;
+
+		int c;
+		while (currentNode != NIL_NODE) {
+			nextNode = helpGetNextNode(currentNode, 0);
+
+			boolean isRemoved = isNodeRemoved(currentNode);
+			if (isRemoved) {
+				// remove the node physically when there is no snapshot running
+				if (highestRequiredSnapshotVersion == 0 && deleteCount < numKeysToDeleteOneTime) {
+					doPhysicalRemove(currentNode, prevNode, nextNode);
+					logicallyRemovedNodes.remove(currentNode);
+					totalSize--;
+					deleteCount++;
+				} else {
+					prevNode = currentNode;
+				}
+				currentNode = nextNode;
+				continue;
+			}
+
+			c = compareByteBufferAndNode(keyByteBuffer, keyOffset, keyLen, currentNode);
+
+			// find the key
+			if (c == 0) {
+				return currentNode;
+			}
+
+			// the key is less than the current node, and nodes after current
+			// node can not be equal to the key.
+			if (c < 0) {
+				break;
+			}
+
+			prevNode = currentNode;
+			currentNode = helpGetNextNode(currentNode, 0);
+		}
+
+		return NIL_NODE;
+	}
+
+	/**
+	 * Put the key into the skip list. If the key does not exist before, a new node
+	 * will be created. If the key exists before, the old state will be returned.
+	 *
+	 * @param keyByteBuffer  byte buffer storing the key.
+	 * @param keyOffset      offset of the key.
+	 * @param keyLen         length of the key.
+	 * @param value          the value.
+	 * @param returnOldState whether to return old state.
+	 * @return the old state. Null will be returned if key does not exist or returnOldState is false.
+	 */
+	@VisibleForTesting
+	S putNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen, byte[] value, boolean returnOldState) {
+		int deleteCount = 0;
+		long prevNode = findPredecessor(keyByteBuffer, keyOffset, 1);
+		long currentNode = helpGetNextNode(prevNode, 0);
+		long nextNode;
+
+		int c;
+		for ( ; ; ) {
+			if (currentNode != NIL_NODE) {
+				nextNode = helpGetNextNode(currentNode, 0);
+
+				boolean isRemoved = isNodeRemoved(currentNode);
+
+				// note that the key may be put again before the node is physically removed,
+				// so the node still need to compare with the key although they can not be
+				// physically removed here
+				if (isRemoved && highestRequiredSnapshotVersion == 0 && deleteCount < numKeysToDeleteOneTime) {
+					doPhysicalRemove(currentNode, prevNode, nextNode);
+					logicallyRemovedNodes.remove(currentNode);
+					totalSize--;
+					deleteCount++;
+					currentNode = nextNode;
+					continue;
+				}
+
+				c = compareByteBufferAndNode(keyByteBuffer, keyOffset, keyLen, currentNode);
+
+				if (c > 0) {
+					prevNode = currentNode;
+					currentNode = nextNode;
+					continue;
+				}
+
+				if (c == 0) {
+					int version = SkipListUtils.helpGetNodeLatestVersion(currentNode, spaceAllocator);
+					boolean needCopyOnWrite = version < highestRequiredSnapshotVersion;
```

Comment:
`version == highestRequiredSnapshotVersion` means this node is put after the latest snapshot get triggered, thus not referenced by any of ongoing snapshots and don't need to do copy on write.

Reply:
Should we rename `highestRequiredSnapshotVersion` then? It is kind of misleading because what it is is actually `highestRequiredSnapsthoVersionPlusOne`.